### PR TITLE
[SERVICES-2436] Filtered staking farms query

### PR DESCRIPTION
**Pull request by mad2sm0key**
## Reasoning
- provide a [Relay](https://relay.dev/)-compliant query for fetching staking farms data

  
## Proposed Changes
- create new `filteredStakingFarms` query that follows the GraphQL "Connections" specification
- add staking farms filtering service (allows wildcard filtering by a token - needed for quick search)
- add method for computing a staking contracts `deployedAt` date (using API service)
- add sorting by farming token price, TVL, APR and deployment date

## How to test
```
query {
  filteredStakingFarms(
    filters:{
      searchToken:"UTK"
    }
    pagination:{
      first:200
    }
    sorting: {
      sortField: TVL
      sortOrder:DESC
    }
  ) {
    edges {
      cursor
      node {
        address
        apr
        state
        farmTokenSupply
        farmingToken {
          identifier
          price
          supply
        }
        farmToken {
          ticker
        }
      }
    }
    pageInfo {
      startCursor
      endCursor
    }
    pageData {
      count
      limit
      offset
    }
  }
}
```

### CODE REVIEW

File: src/modules/staking/services/staking.filtering.service.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        if (
+            !stakingFilter.searchToken ||
+            stakingFilter.searchToken.trim().length < 3
+        ) {
+            return stakeAddresses;
+        }
+
+        const searchTerm = stakingFilter.searchToken.toUpperCase().trim();
+
+        const farmTokens = await Promise.all(
```
Comment (claudiulataretu) [inline src/modules/staking/services/staking.filtering.service.ts:25]: I would keep only the `farmingToken` for this filtering

File: src/modules/staking/services/staking.compute.service.ts [hunk 2]
```diff
@@ -144,6 +146,13 @@ export class StakingComputeService {
         return extraRewardsAPRBoundedPerBlock.multipliedBy(blockDifferenceBig);
     }
 
+    async computeFarmingTokenPriceUSD(stakeAddress: string): Promise<string> {
```
Comment (claudiulataretu) [inline src/modules/staking/services/staking.compute.service.ts:149]: maybe rename it into `farmingTokenPriceUSD`. There is no compute in this method

File: src/modules/staking/services/staking.compute.service.ts [hunk 3]
```diff
@@ -316,4 +325,23 @@ export class StakingComputeService {
...
+
+    @ErrorLoggerAsync({
+        logArgs: true,
+    })
+    @GetOrSetCache({
+        baseKey: 'stake',
+        remoteTtl: CacheTtlInfo.ContractState.remoteTtl,
+        localTtl: CacheTtlInfo.ContractState.localTtl,
+    })
+    async deployedAt(stakeAddress: string): Promise<number> {
```
Comment (claudiulataretu) [inline src/modules/staking/services/staking.compute.service.ts:337]: You can add this as a field into staking farm model and expose it

File: src/modules/staking/services/staking.filtering.service.ts (hunk 1 above)
Comment (mad2sm0key) [reply src/modules/staking/services/staking.filtering.service.ts]: done

File: src/modules/staking/services/staking.compute.service.ts (hunk 3 above)
Comment (mad2sm0key) [reply src/modules/staking/services/staking.compute.service.ts:337]: done

**Review by claudiulataretu — APPROVED** [review]

File: src/modules/staking/services/staking.filtering.service.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        const filteredAddresses: string[] = [];
+        for (const [index, address] of stakeAddresses.entries()) {
+            const farmingToken = farmingTokens[index];
+
+            if (
+                farmingToken.name.toUpperCase().includes(searchTerm) ||
+                farmingToken.identifier.toUpperCase().includes(searchTerm) ||
+                farmingToken.ticker.toUpperCase().includes(searchTerm)
+            ) {
+                filteredAddresses.push(address);
```
Comment (bogdan-rosianu) [inline src/modules/staking/services/staking.filtering.service.ts:40]: not sure, but I think that the above condition will also return VEGLD addresses when searching for EGLD

Comment (mad2sm0key) [reply src/modules/staking/services/staking.filtering.service.ts:40]: That's right. I opted to use `includes` over `startsWith` to account for user typos (say a user misses the E in EGLD, I would still like to show some results). The search is supposed to be as loose as possible.

**Pull request merged**